pkg/clusterstack/factory.go

```diff
@@ -30,21 +28,27 @@ type Printer interface {
 }
 
 type Factory struct {
-	Printer       Printer
-	Fetcher       ImageFetcher
-	Relocator     ImageRelocator
-	TLSConfig     registry.TLSConfig
-	Repository    string
-	BuildImageRef string
-	RunImageRef   string
+	Uploader     Uploader
+	Printer      Printer
+	TLSConfig    registry.TLSConfig
+	Repository   string
+	ValidateOnly bool
 }
 
-func (f *Factory) MakeStack(name string) (*v1alpha1.ClusterStack, error) {
-	if err := f.validate(); err != nil {
+func (f *Factory) MakeStack(name, buildImageTag, runImageTag string) (*v1alpha1.ClusterStack, error) {
+	stackID, err := f.validate(buildImageTag, runImageTag)
+	if err != nil {
 		return nil, err
 	}
 
-	relocatedBuildImageRef, relocatedRunImageRef, stackId, err := f.relocateStack()
+	if f.ValidateOnly {
+		return nil, nil
+	}
```

Comment:
Everything I already stated in my [previous comment](https://github.com/vmware-tanzu/kpack-cli/pull/105#issuecomment-712414476), is still relevant

short-circuiting of execution flow based on a validate defeats the purpose of a `--dry-run`.
the intent of a dry-run is to go through as much as possible of the regular execution flow while skipping state changes.

I intend to make a PR which *builds on* all the refactoring you have done. with both --dry-run and --dry-run-with-image-upload. 
This can be after this PR is merged or before..

Reply:
@sukhil-suresh 

I'm still not clear on what validations are missing and if they are missing, we should add them back!

One thing I like is that we do validations in the factory's `validate` method for clarity. If we are missing any, we should put them there.